runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +150,146 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
 	return count;
 }
 
+/**
+ * Return a String object containing a summary of the classes on the heap, 
+ * the number of instances, and their aggregate size.
+ * This string inserts Unix-style line separators.  The caller is responsible for translating them if necessary.
+ */
+jstring JNICALL
+Java_openj9_tools_attach_diagnostics_base_DiagnosticUtils_getHeapClassStatisticsImpl(JNIEnv * env, jclass unused)
+{
+	J9VMThread *vmThread = (J9VMThread *) env;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9HashTable *statsTable = NULL;
+	J9HashTableState hashTableState;
+	BOOLEAN outOfMemory = FALSE;
+	j9object_t stringObject = NULL;
+	J9HeapStatisticsTableEntry **statsArray = NULL;
+	jstring stringObjectRef = NULL;
+	I_32 numClasses = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	vmFuncs->internalEnterVMFromJNI(vmThread);
+
+	vmFuncs->acquireExclusiveVMAccess(vmThread);
+	statsTable = collectHeapStatistics(vmThread);
+	vmFuncs->releaseExclusiveVMAccess(vmThread);
+
+	if (NULL != statsTable) {
+		numClasses = hashTableGetCount(statsTable);
+		statsArray = j9mem_allocate_memory(numClasses * sizeof(J9HeapStatisticsTableEntry*), J9MEM_CATEGORY_VM_JCL);
+	}
+	if (NULL == statsArray) {
+		outOfMemory = TRUE;
+	} else {
+		I_32 cursor = 0;
+		J9HeapStatisticsTableEntry *entry = (J9HeapStatisticsTableEntry *) hashTableStartDo(statsTable, &hashTableState);
+		while (NULL != entry) {
+			entry->aggregateSize = entry->objectSize * entry->objectCount;
+			statsArray[cursor] = entry;
+			cursor += 1;
+			entry = (J9HeapStatisticsTableEntry *) hashTableNextDo(&hashTableState);
+		}
+		hashTableFree(statsTable);
+		numClasses = cursor; /* adjust the length in case the hash table contained nulls */
+		sortStatsTable(statsArray, 0, numClasses - 1);
+	}
+	if (!outOfMemory) {
```

Comment:
This block could be merged with the `else` block above (there's no need to check `outOfMemory` - it isn't set in the block above). The two following local variables will need to move up.

Reply:
Good idea.